### ISSUE TITLE
New lint that suggests using `addr_of!` instead of `&x as *const _`

### ISSUE THREAD

**Issue by sdroege** [rust-lang/rust-clippy#10872]
### What it does

There's quite a bit of Rust code out there that does the following pattern

```rust
let ptr = &x as *const T;
```

It would seem useful to have a lint that suggests using `addr_of!` (and `addr_of_mut!` for the mutable case) instead.

It should cover the simple case above as well as more complex cases like `&(*some_ptr).x as *const T`, `&x.y as *const T`, `&x.y` (implicit conversion from reference to pointer).

There's also a slightly more convoluted case that would be useful to cover: `&(x.y as *const T) as *const U` (`x.y` is a pointer or reference). In this case it should suggest to also get rid of the inner cast. The inner cast plus `&` here will cause a reference to the local stack frame to be created, which is likely unwanted and likely going to cause unsoundness later on.

That more convoluted case is actually something I had to debug a while ago and it was not easy to spot, and only caused problems in debug builds for extra fun.

### Advantage

The main advantage here is that it creates the pointer without creating a temporary reference first (see `addr_of!` docs). The second more convoluted case would also avoid potential unsoundness.

### Drawbacks

The only possible drawback I can think of is false positives, but the risk seems quite low here.

### Example

```rust
let ptr = &x as *const T;
let ptr2 = &(ptr as *const U) as *const V;
```

Could be written as:

```rust
let ptr = addr_of!(x);
let ptr2 = addr_of!(ptr) as *const V;
```

**Comment by GuillaumeGomez** [comment]
I'll take a look at this in the next days because it seems to be strongly needed.

**Comment by Alexendoo** [comment]
The second case deserves its own lint if we can nail it down, the general `addr_of` suggestion I would imagine would be allow by default if it's linting everywhere it could be used

**Comment by KisaragiEffective** [comment]
I'd like to note that `ptr2` in former example has cast of `&'_ *const U` into `*const V` which is may not intended that has same effect as latter example one.

**Comment by sdroege** [comment]
Yeah I think we should only focus on the first case here. I think I got the essence of the second case now and how that could be turned in a more generic, separate lint (will create an issue in a bit).
